compaction: wait until cg gate is closed before removing compaction s…

Deexie opened the pull request:
…tate

In compaction_group::stop compaction group's gate close() is awaited after all other compaction info is removed. Due to that an operation that entered cg's gate may attempt to access other compaction group related data that was already removed.

Await for cg's closed gate future before associated compaction_state is erased from compaction manager.

Fixes: #19873.

raphaelsc commented:
I think this is fixing the symptom, not the root cause. We're only accessing compaction manager for a group that is removed, because of https://github.com/scylladb/scylladb/issues/19378. After we fix it, we will no longer have table::do_add_sstable_and_update_cache racing with cleanup.

Maybe we have a hidden issue in compaction_manager::remove() since it's supposed to wait for all background tasks (the ones that are ongoing and also that ones that haven't initiated yet). And once compaction_group::gate has close() called on it, no further compactions can be submitted through a compaction_group (even before future of gate::close() hasn't been resolved). Do you think we have a problem in this area? Maybe a task is registered into task manager, but gate in compaction manager for the state wasn't held yet?

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10670/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10670/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10861/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2840/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10670/)
#### Build Details:
- Duration: 4 hr 3 min
- Builder: i-099fbff40eba5dabf (m5d.8xlarge)

raphaelsc commented:
I don't understand exactly what happened yet. I mean, the conditions. For example, if it's caused by a race between split vs migration. It seems to be it, but it would be good to understand why it happened.

Deexie commented:
> Maybe we have a hidden issue in compaction_manager::remove() since it's supposed to wait for all background tasks (the ones that are ongoing and also that ones that haven't initiated yet). And once compaction_group::gate has close() called on it, no further compactions can be submitted through a compaction_group (even before future of gate::close() hasn't been resolved). Do you think we have a problem in this area? Maybe a task is registered into task manager, but gate in compaction manager for the state wasn't held yet?

I do not think task manager has anything to do with that. I looked at gates of `compaction_group` and `compaction_state`.
`compaction_state`'s gate is entered before each compaction executor is created and kept until it's done. `compaction_group`'s gate is held for some compaction types and some operations that don't have their own executor, like flush.

`compaction_group::stop()` does the following:
```
close cg gate
close c_state gate
stop executors
wait for c_state gate
(remove c_state)
wait for cg gate
```

So, I think that the split (or any type of compaction) wasn't the part of the race (it could access compaction_state as it kept its gate and stop would wait for it).

I went through many code paths related to `external_updater` and `compaction_manager::get_compaction_state` (the error is due to get_compaction_state called in external_updater::execute()) but I have no idea what could have happened.

While going through the code I encountered some issues:
- if compaction enters compaction_state gate, yields; compaction_group is stopped and so are ongoing compactions and `stop` waits until compaction_state's gate is closed (yields); we got back to compaction, create an executor - and the executor isn't aborted so cleanup needs to wait until it's done.
- tablet cleanup stops ongoing compactions with "table removal" reason.
- @raphaelsc did you not keep split task in task manager on purpose?

raphaelsc commented:
To add some more info, split compaction always happen in main compaction group, and spans main, left, and right, all of which belong to same storage_group. I think previously, before my rwlock series, we always stopped first the main group, so we didn't face this problem. We can check.

raphaelsc commented:
using tablet shuffle API, I was able to reproduce it

See that it indeed happening when adjust the backlog tracker:

```
INFO  2024-08-01 14:15:09,982 [shard  3:strm] table - adjusting backlog tracker for group 6
ERROR 2024-08-01 14:15:09,983 [shard  3:strm] cache - Fatal error during cache update: std::out_of_range (Compaction state for table [0x60c00505c2e0] not found), at: /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x554d8e /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x5553b0 /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x5556b8 /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x3a089e /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x3a0a57 0x15b9b60 0x10259c7 0x15b97d8 0x1258fd4 0x15b4a90 0x15b47fd 0x138daa5 0x13f9700 0x18c33a6 0x18d5c27 0x184506b 0x1843023 0x183d4e5 /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x453c23
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<sstables::compaction_result>, seastar::noncopyable_function<seastar::future<sstables::compaction_result> ()>, seastar::future<void>::then_impl_nrvo<seastar::noncopyable_function<seastar::future<sstables::compaction_result> ()>, seastar::future<sstables::compaction_result> >(seastar::noncopyable_function<seastar::future<sstables::compaction_result> ()>&&)::{lambda(seastar::internal::promise_base_with_type<sstables::compaction_result>&&, seastar::noncopyable_function<seastar::future<sstables::compaction_result> ()>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<sstables::compaction_result>, seastar::noncopyable_function<seastar::future<sstables::compaction_result> (seastar::future<sstables::compaction_result>&&)>, seastar::future<sstables::compaction_result>::then_wrapped_nrvo<seastar::future<sstables::compaction_result>, seastar::noncopyable_function<seastar::future<sstables::compaction_result> (seastar::future<sstables::compaction_result>&&)> >(seastar::noncopyable_function<seastar::future<sstables::compaction_result> (seastar::future<sstables::compaction_result>&&)>&&)::{lambda(seastar::internal::promise_base_with_type<sstables::compaction_result>&&, seastar::noncopyable_function<seastar::future<sstables::compaction_result> (seastar::future<sstables::compaction_result>&&)>&, seastar::future_state<sstables::compaction_result>&&)#1}, sstables::compaction_result>
   --------
   seastar::internal::coroutine_traits_base<sstables::compaction_result>::promise_type
   --------
   seastar::internal::coroutine_traits_base<sstables::compaction_result>::promise_type
   --------
   seastar::internal::coroutine_traits_base<sstables::compaction_result>::promise_type
   --------
   seastar::internal::coroutine_traits_base<std::optional<sstables::compaction_stats> >::promise_type
   --------
   seastar::shared_future<std::optional<sstables::compaction_stats> >::shared_state
Aborting on shard 3, in scheduling group streaming.
Backtrace:
  /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x4153b4
  /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x3de565
  /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x4095c6
  /lib64/libc.so.6+0x40cff
  /lib64/libc.so.6+0x994a3
  /lib64/libc.so.6+0x40c4d
  /lib64/libc.so.6+0x28901
  /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x3a0a5c
  0x15b9b60
  0x10259c7
  0x15b97d8
  0x1258fd4
  0x15b4a90
  0x15b47fd
  0x138daa5
  0x13f9700
  0x18c33a6
  0x18d5c27
  0x184506b
  0x1843023
  0x183d4e5
  /home/raphaelsc/scylla/build/dev/seastar/libseastar.so+0x453c23
Aborted (core dumped)
```

raphaelsc commented:
Created this PR: https://github.com/scylladb/scylladb/pull/19987.

Deexie commented:
> > While going through the code I encountered some issues:
> > 
> > * if compaction enters compaction_state gate, yields; compaction_group is stopped and so are ongoing compactions and `stop` waits until compaction_state's gate is closed (yields); we got back to compaction, create an executor - and the executor isn't aborted so cleanup needs to wait until it's done.
> 
> compaction_manager::remove() should wait for it, since compaction_state gate is held throughout a compaction task lifetime.

Yes, it will. I meant that it will unnecessarily wait for the whole compaction to finish instead of being aborted and finishing asap. I think we can check if the gate isn't closed before creating the executor.

> 
> > * tablet cleanup stops ongoing compactions with "table removal" reason.
> > * @raphaelsc did you not keep split task in task manager on purpose?
> 
> It reuses same path for executing task. I only mentioned task manager because of my hypothesis that compaction_state's gate was called after yielding, but that's not true I think.

Yes, but `perform_compaction` doesn't keep the executor in task manager if `parent_info == std::nullopt`. This is the leftover from when some compaction types where integrated with task manager and some weren't, which I forgot to remove. I wanted to remove it and checked whether no one uses the no-task-manager option, and split actually does. But after your response I assume I'm free to change that.

```
    // parent_info set to std::nullopt means that task manager should not register this task executor.
    // To create a task manager task with no parent, parent_info argument should contain empty task_info.
    template<typename TaskExecutor, typename... Args>
    requires std::is_base_of_v<compaction_task_executor, TaskExecutor> &&
            std::is_base_of_v<compaction_task_impl, TaskExecutor> &&
    requires (compaction_manager& cm, throw_if_stopping do_throw_if_stopping, Args&&... args) {
        {TaskExecutor(cm, do_throw_if_stopping, std::forward<Args>(args)...)} -> std::same_as<TaskExecutor>;
    }
    future<compaction_manager::compaction_stats_opt> perform_compaction(throw_if_stopping do_throw_if_stopping, std::optional<tasks::task_info> parent_info, Args&&... args);
```

> 
> If we're getting this failure while update_cache, it means it's either memtable flush, or compaction, or some external sstable writer e.g. from repair.
> 
> We don't submit new regular compaction after we closed compaction group's gate (even before future resolves).
> 
> If you look at the error, update_cache happens in streaming sched group.
> 
> I tend to think it's either memtable flush, or some external sstable writer, that adds a sstable after compaction group was stopped. It's not clear from the stack traces.
> 
> Split compaction happens in streaming sched group, and it was happening while we got the error.
> 
> Split compaction spans 3 compaction groups of a given table. It happens on main group, and will add output sstables to left and right groups.
> 
> What if we there's a race where left and right groups are stopped first, not main group. Split completes, calls compaction_group::update_main_sstable_list_on_compaction_completion(), which tries to add sstable to left or right compaction group. See compaction_group::update_main_sstable_list_on_compaction_completion(). It calls backlog_tracker_adjust_charges(), which in turn calls compaction manager for updating compaction_state of left and right groups, but either / both can be stopped already and will trigger an exception, which results in fatal error, since it happens in row_cache::external_updater_impl::execute(). Makes sense to you?

Yeah, absolutely. Nice! I was excluding this part automatically because it was under main cg state gate...
> 
> A possible fix is to patch storage_group::stop() to always close main compaction group first.

denesb commented:
@Deexie what is the state of this PR?

Deexie commented:
> @Deexie what is the state of this PR?

Closing, issue is fixed already
